feat(AIP-140): allow url field name in specific cases

stephanemoore opened the pull request:
This proposes to update the guidelines to advocate for clearer naming for fields that can only return URLs.

It appears to be possible to misinterpret AIP-0140 guidelines to infer that fields that return URLs MUST be named as if they could be URIs. It seems more accurate to indicate that fields that COULD be URIs should use `uri` in the name and fields that MUST be URLs should use `url` in the name.

stephanemoore committed:
Update guidance on URL/URI naming in AIP-140 ✍️

This proposes to update the guidelines to advocate for clearer naming for fields that can only return URLs.

It appears to be possible to misinterpret AIP-0140 guidelines to infer that fields that return URLs MUST be named as if they could be URIs. It seems more accurate to indicate that fields that COULD be URIs should use `uri` in the name and fields that MUST be URLs should use `url` in the name.

noahdietz commented on aip/general/0140.md:
```diff
@@ -137,9 +137,14 @@ this case, services **may** use a `string`.
+
+### URLs
+
+If a field can only represent a URL, the field name should always use `url`.
```
Bold the requirement word
```suggestion
If a field can only represent a URL, the field name **should** always use `url`.
```

stephanemoore replied:
Good idea. Done!

noahdietz commented on aip/general/0140.md:
```diff
@@ -137,9 +137,14 @@ this case, services **may** use a `string`.
+
+### URLs
+
+If a field can only represent a URL, the field name should always use `url`.
```
Would it be reasonable to also ask for some documentation e.g. "Such fields **must** include a comment that documents this format requirement." ? An example proto to go with it would help drive better outcomes because 1. people love copy+pasting 2. AI models will be able to provide more consistent suggestions.

stephanemoore replied:
> Would it be reasonable to also ask for some documentation e.g. "Such fields must include a comment that documents this format requirement." ?

I could not immediately decide whether it is always reasonable to ask for some documentation 🤔

Consider the following example:
```proto
message Book {
  string first_image_url = 1;

  // A URL pointing to an image.
  string second_image_url = 2;

  // A URL pointing to an image.
  // This must always be a URL.
  string third_image_url = 3;
}
```

Which of these field declarations do you think would be acceptable?

`third_image_url` feels a bit redundant to me.
`second_image_url` is what I would typically feel comfortable with.
`first_image_url` feels like it lacks documentation to me.

Thoughts?

> An example proto to go with it would help drive better outcomes

I immediately agree that an example proto would be helpful and have added one 👍

While adding the example proto, I felt like it would be helpful to have an example with both URL and URI fields. In the process of adding that example, I felt that it was relevant to both the existing URIs section and the newly proposed URLs section and decided to explore merging the two sections together. Let me know how this new structure feels and if you think a different structure would be better 🙏

> 1. people love copy+pasting

This is so true 😂🍝

> 2. AI models will be able to provide more consistent suggestions.

Good point!

noahdietz replied:
> `second_image_url` is what I would typically feel comfortable with.

Yeah I agree, I think that's what I would expect at a bare minimum. It implies what (3) explicitly states (+1 to your redundant comment) and if a more specific comment than this is necessary, it can be recommended/added.

> I felt that it was relevant to both the existing URIs section and the newly proposed URLs section and decided to explore merging the two sections together.

I actually like this a lot, great adjustment.

noahdietz reviewed (commented):
Hey @stephanemoore this seems reasonable to do! Let me circle back with some others and get back to you.

noahdietz commented on aip/general/0140.md:
```diff
@@ -137,9 +137,14 @@ this case, services **may** use a `string`.
+
+Field names **may** use a prefix in front of `uri` as appropriate.
+
+### URLs
```
It would also be good for historical context to have a `**Note:**` callout suggesting that APIs that have followed the URIs pattern should continue to do so and not change things unnecessarily or diverge from local consistency.

stephanemoore replied:
I added a note but I hesitated to write "should" and went with "may".

I worry that using "should" could result in APIs being anchored on historical convention and never seeking a path towards following latest recommendations. For example, a team could choose to use the latest conventions in newly added protos but I feel like they would be less likely to do so if historical inertia for the overall API carries too much weight. (I actually just had such a conversation in an API review.)

noahdietz replied:
Great compromise, thank you. SGTM!

noahdietz commented on aip/general/0140.md:
```diff
@@ -137,9 +137,14 @@ this case, services **may** use a `string`.
+
+### URLs
+
+If a field can only represent a URL, the field name should always use `url`.
```
```suggestion
If a field can only represent a URL, the field name should use `url`.
```
drop "always" which conflicts with "should" and leaks into "must" territory

stephanemoore committed:
Apply suggestion to drop "always"

Co-authored-by: Noah Dietz <noahdietz@users.noreply.github.com>

stephanemoore committed:
Apply various changes based on review feedback in PR #1551

Changes include:
* Remove "always" from the prescription for URI fields for
  consistency.
* Collapse proposed URLs section into URIs section.
* Reformat prescription for URIs to be consistent with prescription for
  URLs.
* Add an example proto for clarity.
* Add a note indicating that the prescription for URLs should be
  balanced against local consistency and API stability.

PR link:
https://github.com/aip-dev/google.aip.dev/pull/1551

noahdietz commented on aip/general/0140.md:
````diff
@@ -137,14 +137,29 @@ this case, services **may** use a `string`.
+```
+
+**Note:**
+APIs that have previously used `uri` for URL fields may continue to do so
````
```suggestion
**Note:** APIs that have previously used `uri` for URL fields may continue to do so
```

Nit: Put them on the same line

stephanemoore replied:
Collapsed them onto the same line and wrapped to 80 columns.

stephanemoore committed:
Remove newline after `Note:` in URIs section
